No special treatment of pre blocks in loops

nanspro opened the pull request:
### Description

I have removed the steps where handling of pre blocks of for loops was done before, instead of that an assert statement is there which checks if a pre block for loops are empty or not.

Fixes #6202 

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] Used meaningful commit messages

nanspro committed:
Removing scope rules for pre block in 'For Loops'

stackenbotten commented:
There was an error when running `test_check_style` for commit `b2d05d2cbaa2b90ce4f0b24eb28c01f704ef51b9`:
```
Error: Trailing whitespace found:
 test/libyul/yulOptimizerTests/expressionSplitter/control_flow.yul:4: let a := 2 
 test/libyul/yulOptimizerTests/redundantAssignEliminator/for.yul:4: for { 

```
Please check that your changes are working as intended.

stackenbotten commented:
There was an error when running `test_check_style` for commit `b8e869c8edb49bc82a98b599e9825dbbaab43888`:
```
Error: Trailing whitespace found:
 libyul/optimiser/ExpressionSplitter.cpp:70:{ 
 libyul/optimiser/ExpressionSplitter.cpp:72: 
 test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul:8: for { 

```
Please check that your changes are working as intended.

nanspro commented:
@chriseth  could you review?
I need help with some tests, you can see i changed `SSATransform` and then i changed the tests `ssaTransform` and it passed but when i changed the for loops in `ssaAndBack` it is giving an error

```
unknown location(0): fatal error: in "yulOptimizerTests/ssaAndBack/for_loop": signal: SIGSEGV, si_code: 0 (memory access violation at address: 0x00000080)
```
A similar thing happens with `ExpressionSplitter` the tests `expressionSplitter` are passing but `expressionSimplifier` are failing with error
```
unknown location(0): fatal error: in "yulOptimizerTests/expressionSimplifier/inside_for": memory access violation at address: 0x7f8d0000001a: no mapping at fault address
```
What does these error means, can't seem to find anything anywhere?

Marenz commented:
fyi, chris is on vacation the next few days

jjlk40 commented:
Jason Lake has sent you an email via Gmail confidential mode:Re: [ethereum/solidity] No special treatment of pre blocks in loops (#6486)This message was sent on Apr 8, 2019 at 1:04:51 PM PDTYou can open it by clicking the link below. This link will only work for reply@reply.github.com.View the emailGmail confidential mode gives you more control over the messages you send. Set an expiration time, disable printing or forwarding of a message and more. Learn moreGmail: Email by GoogleUse is subject to the Google Privacy PolicyGoogle LLC, 1600 Amphitheatre Parkway, Mountain View, CA 94043, USAYou have received this message because someone sent you an email via Gmail confidential mode.

nanspro commented:
@Marenz  okay, if anyone gets time and can help me a little bit with these tests it'd be great. This is my first contribution to solidity and i can't seem to find the test errors reason anywhere.

Marenz commented:
So the errors you posted above suggest some memory management problems. I would recommend to run the tests with gdb, that should show you which variable accesses/functions are crashing and from there you can track how the variable is pointing to invalid memory.

Be sure to build with debug symbols before you do this

ekpyron commented on libyul/optimiser/DataFlowAnalyzer.cpp:
```diff
@@ -112,10 +112,8 @@ void DataFlowAnalyzer::operator()(FunctionDefinition& _fun)
 void DataFlowAnalyzer::operator()(ForLoop& _for)
 {
-	// Special scope handling of the pre block.
-	pushScope(false);
```
You removed the ``pushScope`` here, but still call ``popScope`` below - my guess is that that's the problem.
The fix is probably to just leave the ``pushScope(false);`` in here.

Marenz commented on test/libyul/yulOptimizerTests/expressionSimplifier/inside_for.yul:
```diff
@@ -1,12 +1,13 @@
 {
-    for { let a := 10 } iszero(eq(a, 0)) { a := add(a, 1) } {}
+    let a := 10
```
I am not very familiar with the ticket you're solving, but I believe that this should still work after the changes. @ekpyron confirm please? (ticket in question is #6202)

Marenz replied:
Ah, but because that is done in a different optimization step it won't work in the individual optimization tests of other steps.. that could make sense..

ekpyron replied:
Well, it depends :-). The ticket is meant to remove special treatment of pre-blocks in all optimizer components and rather let them depend on the fact that pre blocks were already removed before in the ``ForLoopInitRewriter``.
The tests that only call one isolated optimizer component need to deal with that dependency - so either we don't use pre-blocks in such tests (that is this change here) or we call the ``ForLoopInitRewriter`` before running such a test run in ``YulOptimizerTest.cpp``. Both options are fine and I wouldn't prefer either, so we can leave it at this.

ekpyron replied:
In short: It's fine as it is ;-).

nanspro commented:
@Marenz  @ekpyron  FIXED!!

ekpyron commented on test/libyul/yulOptimizerTests/fullInliner/not_inside_for.yul:
```diff
@@ -13,12 +14,12 @@
 // ----
 // {
 //     {
+        let a_3 := 0
```
I'm surprised that this works, since it's missing the ``//`` at the line beginning - it seems they are in fact optional, but you should still add them for consistency. The same applies to all the other test expectations (everything below the ``// ====`` should start with ``//``.

nanspro replied:
Done.

ekpyron reviewed (changes requested):
This looks almost good now! Can you address the remaining style issues and squash the commits into a single commit?

leonardoalt reviewed (changes requested):
In the header files of the modified classes you will find a line
`Prerequisite: ...`
`ForLoopInitRewriter` needs to be added there for each of them.

After you fix the indentation and comments suggestions from @ekpyron about the tests, could you please squash?

chriseth commented on libyul/optimiser/DataFlowAnalyzer.cpp:
```diff
@@ -112,10 +112,9 @@ void DataFlowAnalyzer::operator()(FunctionDefinition& _fun)
 void DataFlowAnalyzer::operator()(ForLoop& _for)
 {
-	// Special scope handling of the pre block.
 	pushScope(false);
```
The `pushScope` / `popScope` pair here can be removed.

chriseth commented on libyul/optimiser/DataFlowAnalyzer.cpp:
```diff
@@ -112,10 +112,9 @@ void DataFlowAnalyzer::operator()(FunctionDefinition& _fun)
 	pushScope(false);
-	for (auto& statement: _for.pre.statements)
-		visit(statement);
+	// empty pre block in loops
```
Please add a comment: `If the pre block was not empty, we would have to deal with more complicated scoping rules.`

chriseth commented on libyul/optimiser/ExpressionSplitter.cpp:
```diff
@@ -68,7 +68,7 @@ void ExpressionSplitter::operator()(Switch& _switch)
 void ExpressionSplitter::operator()(ForLoop& _loop)
 {
-	(*this)(_loop.pre);
+	assertThrow(_loop.pre.statements.empty(), OptimizerException, "");
```
I don't see a reason why the ExpressionSplitter should require empty pre blocks (please also change the header file and remove the requirement).

chriseth commented on libyul/optimiser/RedundantAssignEliminator.cpp:
```diff
@@ -121,9 +121,7 @@ void RedundantAssignEliminator::operator()(ForLoop const& _forLoop)
 	set<YulString> outerDeclaredVariables;
 	swap(m_declaredVariables, outerDeclaredVariables);
 
-	// We need to visit the statements directly because of the
```
Please also add a similar comment here.

chriseth commented on libyul/optimiser/RedundantAssignEliminator.cpp:
```diff
@@ -121,9 +121,7 @@ void RedundantAssignEliminator::operator()(ForLoop const& _forLoop)
 	set<YulString> outerDeclaredVariables;
```
All the handling of `m_declaredVariables` in this function can now be removed.

chriseth commented on libyul/optimiser/SSATransform.cpp:
```diff
@@ -40,11 +40,7 @@ void SSATransform::operator()(Identifier& _identifier)
 
 void SSATransform::operator()(ForLoop& _for)
 {
-	// This will clear the current value in case of a reassignment inside the
```
I think it is better to keep the SSATransform handle the pre loop - it is not a complication for it, just an inefficiency. Also ForLoopInitRewriter should not be a requirement for the SSA transform.

chriseth commented:
Please add `Prerequisite: ForLoopInitRewriter.` to all optimizer step classes that inherit from DataFlowAnalyzer.

chriseth commented on test/libyul/yulOptimizerTests/splitJoin/control_flow.yul:
```diff
@@ -1,6 +1,7 @@
 {
     if mul(add(calldataload(0), 2), 3) {
-        for { let a := 2 } lt(a, mload(a)) { a := add(a, mul(a, 2)) } {
```
I think this change can/should be reverted if expressionSplitter does not require ForLoopInitRewriter anymore.

chriseth commented on test/libyul/yulOptimizerTests/splitJoin/functions.yul:
```diff
@@ -3,7 +3,8 @@
     function f(y) -> r {
         r := mload(mul(6, add(y, 0x20)))
     }
-    for { let a := 2 } lt(a, mload(a)) { a := add(a, mul(a, 2)) } {
```
Also here.

chriseth commented on test/libyul/yulOptimizerTests/ssaAndBack/for_loop.yul:
```diff
@@ -1,7 +1,7 @@
 {
+  let a := mload(0)
```
Similarly here.

nanspro replied:
@chriseth  on running this test(after adding pre blocks in loop) i am getting this error( it points to empty pre block asserting statement), does that mean ssaBack tests are run through `RedundantAssignEliminator` first?
```

/home/fnatic/solidity/test/boostTest.cpp(116): error: in "yulOptimizerTests/ssaAndBack/for_loop": Exception during extracted test: /home/fnatic/solidity/libyul/optimiser/RedundantAssignEliminator.cpp(123): Throw in function virtual void yul::RedundantAssignEliminator::operator()(const yul::ForLoop&)

```

chriseth replied:
You are right, these tests have to be modified!

chriseth commented on test/libyul/yulOptimizerTests/ssaTransform/for_reassign_body.yul:
```diff
@@ -1,6 +1,7 @@
 {
     let a := mload(0)
-    for { mstore(0, a) } a { mstore(0, a) }
```
Also here and in all the other ssaTransform tests.

chriseth commented on test/libyul/yulOptimizerTests/expressionSplitter/control_flow.yul:
```diff
@@ -1,7 +1,8 @@
 {
 	let x := calldataload(0)
 	if mul(add(x, 2), 3) {
-		for { let a := 2 } lt(a, mload(a)) { a := add(a, mul(a, 2)) } {
```
I think this test does not have to be modified.

nanspro replied:
Yeah forgot that this doesn't apply to loop conditions

chriseth commented on test/libyul/yulOptimizerTests/fullSimplify/inside_for.yul:
```diff
@@ -1,13 +1,14 @@
 {
     let x := calldataload(3)
-    for { let a := 10 } iszero(eq(a, sub(x, calldataload(3)))) { a := add(a, 1) } {}
```
Does this have to be modified?

nanspro replied:
yes, data analyzer is run upon this one

chriseth commented on test/libyul/yulOptimizerTests/rematerialiser/branches_for_declared_in_init1.yul:
```diff
@@ -1,6 +1,8 @@
 {
     let b := 0
-    for { let a := caller() pop(a) } a { pop(a) } {
```
I think this test can be removed - there is no point anymore.

chriseth commented on test/libyul/yulOptimizerTests/rematerialiser/branches_for_declared_in_init2.yul:
```diff
@@ -1,6 +1,8 @@
 {
     let b := 0
-    for { let a := caller() pop(a) } lt(a, 0) { pop(a) a := add(a, 3) } {
+    let a := caller()
```
I think this test can be removed - there is no point anymore.

chriseth commented:
The following components need a note that they depend on the ForLoopInitRewriter:
 - CommonSubexpressionEliminator
 - DataFlowAnalyzer
 - ExpressionSimplifier
 - RedundantAssignEliminator
 - Rematerializer
 - StructuralSimplifier

And in addition to those, of course the ones you modified.

If you are fed up with this issue, we can also finish it for you :)

nanspro commented:
@chriseth  check it out now, i have added the prerequisites and all the changed tests are run by some class which is dependent on ForLoopInitRewriter so changes are mandatory.

chriseth commented:
Wonderful, thanks a lot!
